File: Part10_Model_Selection_And_Boosting/Section39_XGBoost/xgboost.py
# XGBoost

# Install xgboost following the instructions on this link:
# http://xgboost.readthedocs.io/en/latest/build.html#

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
from sklearn.preprocessing import LabelEncoder  # encoding categorical data
from sklearn.preprocessing import OneHotEncoder  # dummy encoding
# from sklearn.cross_validation import train_test_split  # deprecated
from sklearn.model_selection import train_test_split  # splitting the dataset
from sklearn.metrics import confusion_matrix
from xgboost import XGBClassifier
from sklearn.model_selection import cross_val_score  # k-Fold Cross Validation


# To see the full array, run the following
np.set_printoptions(threshold=np.nan)

# Importing the dataset
os.chdir("C:/Development/Courses/Kirill Eremenko Data Science Courses/\
Machine_Learning_A-Z/Part10_Model_Selection_And_Boosting/\
Section39_XGBoost")
dataset = pd.read_csv("Churn_Modelling.csv")

# Create a matrix of features
X = dataset.iloc[:, 3:13].values  # 13 because upper bound excluded in range

# Create the dependent variable vector; last column only
y = dataset.iloc[:, 13].values

# Ecoding categorical data (country and gender)
# Country first (the independent variable)
labelencoder_X_1 = LabelEncoder()  # country
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])  # country only
labelencoder_X_2 = LabelEncoder()  # gender
X[:, 2] = labelencoder_X_1.fit_transform(X[:, 2])  # country only

# Dummy encoding - 1 column for each country that has 1 or 0
# Column 0 = France, 1 = Germany, 2 = Spain
onehotencoder = OneHotEncoder(categorical_features=[1])
X = onehotencoder.fit_transform(X).toarray()

# Remove first country dummy variable to avoid dummy variable trap
X = X[:, 1:]

# Splitting the dataset into Training set and Test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,
                                                    random_state=0)
# random_state set to 0 so we all get the same result
# 42 is a good choice for random_state otherwise

# Fitting XGBoost to the Training set
classifier = XGBClassifier()
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

# Making the Confusion Matrix
cm = confusion_matrix(y_test, y_pred)

# Applying k-Fold Cross Validation
# 10 elements with 10 accuracies to evaluate the model
accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train,
                             cv=10)
accuracies.mean()
accuracies.std()

File: Part10_Model_Selection_And_Boosting/Section39_XGBoost/xgboost.R
# XGBoost

# Importing the dataset
setwd("C:/Development/Courses/Kirill Eremenko Data Science Courses/Machine_Learning_A-Z/Part10_Model_Selection_And_Boosting/Section39_XGBoost")
dataset <- read.csv("Churn_Modelling.csv")
dataset <- dataset[4:14]

# Encoding the categorical data as factors
# Have to be numeric because of model we'll be using
# Transform Geography to a column of factors
dataset$Geography <- as.numeric(factor(dataset$Geography,
                                       levels=c("France", "Spain", "Germany"),
                                       labels=c(1, 2, 3)))

# Transform Gender to a factor
dataset$Gender <- as.numeric(factor(dataset$Gender,
                                    levels=c("Female", "Male"),
                                    labels=c(1, 2)))

# Splitting the dataset into Training set and Test set
library(caTools)
set.seed(123)  # like random_state
split <- sample.split(dataset$Exited, SplitRatio=0.8)  # for training set
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)

# Fitting XGBoost to the Training set
# install.packages("xgboost")
library(xgboost)
# training_set$Exited is dependent variable as a vector
classifier <- xgboost(data=as.matrix(training_set[-11]), label=training_set$Exited,
                      nrounds=10)

# Predicting the Test set results
y_pred <- predict(classifier, newdata=as.matrix(test_set[-11]))
y_pred <- (y_pred >= 0.5)

# Making the Confusion Matrix
cm = table(test_set[, 11], y_pred)

# Applying k-Fold Cross Validation
library(caret)
folds <- createFolds(training_set$Exited, k=10)
cv <- lapply(folds, function(x) {  # x is each element of folds
  training_fold <- training_set[-x, ]
  test_fold <- training_set[x, ]
  classifier <- xgboost(data=as.matrix(training_set[-11]), label=training_set$Exited,
                        nrounds=10)
  y_pred <- predict(classifier, newdata=as.matrix(test_fold[-11]))
  y_pred <- (y_pred >= 0.5)
  cm <- table(test_fold[, 11], y_pred)
  accuracy <- (cm[1,1] + cm[2,2])/(cm[1,1] + cm[2,2] + cm[1,2] + cm[2,1])
  return(accuracy)
})
accuracy <- mean(as.numeric(cv))
